db/product.go

```diff
@@ -0,0 +1,313 @@
+}
+
+func (product *Product) Validate() (errorResponse map[string]ErrorResponse, valid bool) {
+	fieldErrors := make(map[string]string)
+
+	if product.Name == "" {
+		fieldErrors["product_name"] = "Can't be blank"
+	}
+	if product.Description == "" {
+		fieldErrors["product_description"] = "Can't be blank "
+	}
+	if product.Price <= 0 {
+		fieldErrors["price"] = "Can't be blank  or less than zero"
+	}
+	if product.Discount < 0 {
+		fieldErrors["discount"] = "Can't be less than zero"
+	}
+	if product.Tax < 0 {
+		fieldErrors["tax"] = "Can't be less than zero"
+	}
+	// If Quantity gets's < 0 by UpdateProductStockById Method, this is what saves us
+	if product.Quantity < 0 {
+		fieldErrors["available_quantity"] = "Can't be blank or less than zero"
+	}
+	if product.CategoryId == 0 {
+		fieldErrors["category_id"] = "Can't be blank"
+	}
+
+	if len(fieldErrors) == 0 {
+		valid = true
+		return
+	}
+
+	errorResponse = map[string]ErrorResponse{
+		"error": ErrorResponse{
+			Code:    "Invalid_data",
+			Message: "Please Provide valid Product data",
+			Fields:  fieldErrors,
+		},
+	}
+	// TODO Other Validations
+	return
+}
+
+// @Title GetProductByID
+// @Description Get a Product Object by its Id
+// @Params req.Context, product's Id
+// @Returns Product Object, error if any
+func (s *pgStore) GetProductByID(ctx context.Context, Id int) (product Product, err error) {
+
+	err = s.db.Get(&product, getProductByIDQuery, Id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting product from database by id: " + string(Id))
+		return
+	}
+
+	// Add category to Product's object
+	var category string
+	err = s.db.Get(&category, getCategoryByID, product.CategoryId)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error fetching category from database by product_id: " + string(Id))
+		return
+	}
+
+	product.CategoryName = category
+
+	// Add Product Image URL's to Product's Object
+	productImage, err := s.GetProductImagesByID(ctx, Id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting productImage from database by id " + string(Id))
+		return
+	}
+
+	for j := 0; j < len(productImage); j++ {
+		product.URLs = append(product.URLs, productImage[j].URL)
+	}
+
+	return
+}
+
+// @Title ListProducts
+// @Description Get limited number of Products of particular page
+// @Params req.Context , limit, page
+// @Returns Count of Records, error if any
+func (s *pgStore) ListProducts(ctx context.Context, limit string, page string) (count int, products []Product, err error) {
+
+	resultCount, err := s.db.Query(getProductCount)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error fetching Cunt of Products from database")
+		return
+	}
+
+	for resultCount.Next() {
+		err = resultCount.Scan(&count)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error scanning Count Of Product into integer variable")
+			return
+		}
+	}
+
+	if count == 0 {
```

Comment:
this will be always true right

Reply:
err = result.Scan(&count)  above it, so it won't always return true.